### ISSUE TITLE
Shuffle sides does not respect teams

### ISSUE THREAD

**Issue by Pentarctagon** [wesnoth/wesnoth#4570]
Originally reported [here](https://steamcommunity.com/app/599390/discussions/0/1640927348828386536/).

When the Shuffle Sides checkbox is selected, it can result the sides chosen by the players during game creation being overridden, and also somehow make sides on different teams allies with each other.  For example, a starting configuration like [this](https://i.imgur.com/7wBZNes.png) ends up with a result like [this](https://imgur.com/a/Ma5pzhs), where both the AI and humans are mixed up as well as East and West teams somehow allied.

**Comment by gfgtdf** [comment]
more importantly: why do the team names on the second screen not match which sides you see information for (not sure whether i expressed myself understanable)

**Comment by Pentarctagon** [comment]
If I'm understanding, you're saying the same thing I meant by "East and West teams somehow allied" - you can see the information because (somehow) two sides on team East are allied with a side on team West.

I'd expect that shuffle sides would randomize the starting positions only, but it seems like for some reason wesnoth is then re-applying some of the side's information based on the starting position.

So what I'd expect in this case would be:
* Geographically, the east side of the map would have 2 players on the East team and one player on the West team.  The geographically west side of the map would then have the opposite.
* Everyone on the East team remains allied.  Everyone on the West team remains allied.

Instead for some reason I guess it's using the starting positions to determine who's allied?

**Comment by gfgtdf** [comment]
> I'd expect that shuffle sides would randomize the starting positions only

well if it did only  randomize the starting positions then that'd mean that enemy team could for example share one caste in that 2vs2 map where llied player share a castle.

> wesnoth is then re-applying some of the side's information based on the starting position.

it somewhat has to do that if it wants to stay compatible with certain scenarios, thing of example of scenarios where side 2 get a map advantage (closer village)

**Comment by CelticMinstrel** [comment]
The shuffle sides implementation is pretty complicated and broken in mutiple ways if recall correctly. I wonder if it might be better to just remove it outright.

Then if people _really_ want the feature, we can define exactly how it _should_ work and reimplement it to that specification. It might turn out that there are multiple valid ways, so it would turn into a dropdown option rather than a simple toggle.

**Comment by gfgtdf** [comment]
well, for a basic start we have starting positions, teams, and players, so whsould tems be 'tied' to starting position or to players? I'd say team becasue of what i said above. so basically each side is 'split' into two parts (A and B) and then the bsa are shuffled  and readded to the A's, we have to think about what belong in A and what in B:
my suggestion is
A:
side index
starting position (because of side specific compensation as noted above)
teams (because otherwise we might end up in strange situation where for example  enemies share one castle as mentioned above)
villgaes (same reason as starting positions)
probably everything else.
B:
Players
Factions

**Comment by sevu** [comment]
I took me an hour to figure out how shuffle side works.

**Comment by sevu** [comment]
This explained a lot:
> After shuffling, wesnoth is then re-applying some of the side's information based on the starting position.

Gfgtdf explained why team_name is currently reapplied. Thing is, user_team_name is not reapplied, and this is what causes a lot of confusion. Well, actually, only when use map settings is active, otherwise the game adds it own team names, and they are matching the actual teams.

deafeat_condition is also not reapplied, but should. 5p wild land is an example, where side 5 has no leader.
Can this be changed in 1.14?


Many other things which are shuffled would currently better fit into category A.
Mostly things which don't matter / differ, and in case they do, are rather different because of the position on the map, thus better not shuffled.
_________________________________________________________________________

Should Gold / Income be reapplied?
Yes, because of scenarios like Auction X. (manually fixed there currently)
No, because in the UI before game start one can change gold/income individually for players. People might agree to give some players an advantage because they are less experienced.

Which leads us back to the topic title:

I would argue that anything which can be changed in the pre-game setup (MP staging) should be tied to the player and thus be shuffled.
And Teams can also be changed in the pre-game setup.

___________________________________________________________________________

Is there a comprises possible, and how?
Say we shuffle team names, and want to handle the situation gfgtdf outlined earlier, with 2 allies sharing one castle, where shuffling team_names can put enemy sides in the same castle.

To achieve that,
 -> we could only shuffle the starting positions among team members. Problem: why would you want to shuffle to begin with?
-> could shuffle in two steps:
-----> grouping team One's and Two's staring position. Trowing the coin who which team gets team One's positions, …
-----> and then shuffling the sides.
Problem: only possible if the teams have equally much starting positions. Mostly they have, but the algorithm needs a solution in case that not.

**Comment by gfgtdf** [comment]
> Say we shuffle team names, and want to handle the situation gfgtdf outlined earlier, with 2 allies sharing one castle, where shuffling team_names can put enemy sides in the same castle.

I'm not sure sure here i think there are situations where players actually _want_ shuffle sides to decide their teams. If you have 4 players for a 2vs2 game and cannot decide who teams up with who you can just enable shuffle sides.


Also the names for the teams are often related to the map, like "North" and "South" furtheremor the scenarios wml code might make assumptions on that the team_names actually match what is specified in wml. 


That said, i think the main situation for the "shuffle sides" feature are classic 1vs1 game, While in theory it should'n really matter how we handle teams there, i think it makes sense to keep the name ties to the stating position because as i said the team names are often given by the map like "North" and "South"

**Comment by sevu** [comment]
So, what to do? I understand the reasoning, labeling this as won't fix?

And having a new issue for user_team_name not being handled correctly (should be handled like team_name, category A in above)

The problem in the steam link is more about being aware that knowing that shuffling side was used, maybe some kind of UI indicator. Like, maybe it can be injected into the objectives in the hint line, or such a popup like when picking up an object.
Though, I guess such an idea might never be implemented.

**Comment by MXB2001** [comment]
What actually happens is not that teams are mixed. Shuffle works perfectly fine, preserving alliances correctly.

The bug is cosmetic in nature only.

Example:
4 players
p1 team Defenders
p2 team Defenders
p3 team Attackers
p4 team Attackers

What will happen is that p1 & 2 will be allied and 3/4 also. However the team names (not the alliances just the labels!) will be shuffled along with the player.

Another example:
3 players
p1 team Acheron
p2 team Barbary
p3 team Crioline

if you choose p1 you may get shuffled to side 2 or 3. However, the team Acheron will stay with you.

In summary the team names move when they ought to be fixed.
